### ISSUE TITLE
asymmetric `TypeIs` using intersection doesn't work

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#4090]
is this valid? i think it could work
```py
from typing import Literal, TypeIs

def positive_int(a: object)  -> (Literal[True] & TypeIs[int]) | Literal[False]:
    return isinstance(a, int) and a >= 0  # error: Return type does not match returned value: expected `Literal[False]`, found `bool`

def m(x: int | str):
    if positive_int(x):
        reveal_type(x)  # Never, expect int
    else:
        reveal_type(x)  # int | str
```

[Playground](https://play.ty.dev/3a5da5e7-df5b-46c8-b9c2-2c5a464e5a39)

**Comment by carljm** [comment]
Currently we consider `Literal[True]` and `TypeIs[int]` to be disjoint types. I think we probably shouldn't do that.

Then I guess it wouldn't be too hard to support this, but it would add some additional work in evaluating return types of narrowing predicate functions, as well as additional work in validating the predicate functions themselves. Not clear if this is worth it -- the pattern seems quite hard for a reader to understand. It isn't quite equivalent to `TypeGuard[int]` (intersects prior type instead of replacing it in positive narrowing branch), but it's close.

**Comment by DetachHead** [comment]
Wouldn't it make more sense to just fix `TypeGuard`?

**Comment by carljm** [comment]
I assume by "fix `TypeGuard`" you mean change its positive-branch narrowing semantics to "intersect" rather than "replace"? I guess the question is whether there is sufficient consensus that that's a bug to justify changing it in a backwards-incompatible way.

**Comment by DetachHead** [comment]
i don't see how it would be backwards-incompatible though. it's just making the type narrower, which i can only imagine would reduce the number of false positive errors, right?

**Comment by carljm** [comment]
I guess it could break some `assert_type`.

**Comment by DetachHead** [comment]
that's a risk with basically every change you could make to a type checker, even bug fixes. maybe it's worth at least trying and seeing what mypy_primer says?

**Comment by carljm** [comment]
You can certainly submit a ty PR and our CI will show the ecosystem impact.
